File: haskell/typelayer-c/src/TypelayerC/Schemas/World/Collections.hs
{-# LANGUAGE GADTs, OverloadedStrings, StandaloneDeriving, DeriveAnyClass, GeneralisedNewtypeDeriving #-}
module TypelayerC.Schemas.World.Collections where
import qualified TypelayerC.Schemas.World.Dynamic as TypelayerCSchemasWorldDynamic
import qualified TypelayerC.Schemas.World.Structures as TypelayerCSchemasWorldStructures
import qualified TypelayerCommon.Collections as TypelayerCommonCollections
import qualified TypelayerCommon.Dynamic as TypelayerCommonDynamic
import qualified TypelayerCommon.Prelude as TypelayerCommonPrelude
import qualified TypelayerCommon.Types as TypelayerCommonTypes


data FileKV = 
    FileKV
        { fileKVKey :: TypelayerCommonTypes.Text
        , fileKVValue :: TypelayerCSchemasWorldStructures.File
        }
    deriving (Show, Eq)

class (Show a, Eq a, TypelayerCommonTypes.Typeable a) => FileCollection a where
    fileCollectionDelete :: a -> [TypelayerCommonTypes.Text] {-toDelete-} -> TypelayerCommonPrelude.ExecuteIO [TypelayerCommonTypes.Text]
    fileCollectionFetch :: a -> [TypelayerCommonTypes.Text] {-toFetch-} -> TypelayerCommonPrelude.ExecuteIO [FileKV]
    fileCollectionSave :: a -> [FileKV] {-toSave-} -> TypelayerCommonPrelude.ExecuteIO [TypelayerCommonTypes.Text]
    fileCollectionSlice :: a -> Integer {-offset-} -> Integer {-count-} -> TypelayerCommonPrelude.ExecuteIO [FileKV]


data PackedFileCollection where
    PackedFileCollection :: FileCollection a => a -> PackedFileCollection
deriving instance Show PackedFileCollection
deriving instance TypelayerCommonTypes.Typeable PackedFileCollection
instance Eq PackedFileCollection where
    (==) (PackedFileCollection a) (PackedFileCollection b) = 
      case (TypelayerCommonPrelude.cast a) of
        Nothing -> False
        Just a' -> a' == b

instance FileCollection PackedFileCollection where
    fileCollectionDelete (PackedFileCollection x) toDelete = fileCollectionDelete x toDelete
    fileCollectionFetch (PackedFileCollection x) toFetch = fileCollectionFetch x toFetch
    fileCollectionSave (PackedFileCollection x) toSave = fileCollectionSave x toSave
    fileCollectionSlice (PackedFileCollection x) offset count = fileCollectionSlice x offset count

data FileDynamicCollection = 
    FileDynamicCollection
    { fileDynamicCollectionDynamicObj :: TypelayerCommonCollections.PackedBinaryCollection
    }
    deriving (Show, Eq)


instance FileCollection FileDynamicCollection where
    fileCollectionDelete this toDelete = do
        let dynamicObject = (fileDynamicCollectionDynamicObj this)
        (TypelayerCommonCollections.binaryCollectionDelete dynamicObject toDelete)
    fileCollectionFetch this toFetch = do
        let dynamicObject = (fileDynamicCollectionDynamicObj this)
        dynamicKvs <- (TypelayerCommonCollections.binaryCollectionFetch dynamicObject toFetch)
        TypelayerCommonPrelude.listReduceM dynamicKvs [] $ \reduceDynamicKvsIndex reduceDynamicKvsElem reduceDynamicKvsAcc -> do
            collectionEncoding <- (TypelayerCommonCollections.binaryCollectionEncoding dynamicObject )
            newDecoder <- (TypelayerCommonDynamic.encodingStartDecoding collectionEncoding (TypelayerCommonCollections.binaryKVValue reduceDynamicKvsElem))
            fileValue <- (TypelayerCSchemasWorldDynamic.dynamicGetObjFile newDecoder)
            let appendList0 = ((FileKV{fileKVKey=(TypelayerCommonCollections.binaryKVKey reduceDynamicKvsElem), fileKVValue=fileValue}):reduceDynamicKvsAcc)
            return appendList0
    fileCollectionSave this toSave = do
        let dynamicObject = (fileDynamicCollectionDynamicObj this)
        dynamicList <- do
            TypelayerCommonPrelude.listReduceM toSave [] $ \convertLoopIndex convertLoopElem convertLoopAcc -> do
                collectionEncoding <- (TypelayerCommonCollections.binaryCollectionEncoding dynamicObject )
                newEncoder <- (TypelayerCommonDynamic.encodingStartEncoding collectionEncoding )
                elemEncoded0 <- (TypelayerCSchemasWorldDynamic.dynamicPutFile newEncoder (fileKVValue convertLoopElem))
                elemBinary <- (TypelayerCommonDynamic.encoderToBinary elemEncoded0 )
                let appendList0 = ((TypelayerCommonCollections.BinaryKV{TypelayerCommonCollections.binaryKVKey=(fileKVKey convertLoopElem), TypelayerCommonCollections.binaryKVValue=elemBinary}):convertLoopAcc)
                return appendList0
        (TypelayerCommonCollections.binaryCollectionSave dynamicObject (reverse dynamicList))
    fileCollectionSlice this offset size = do
        let dynamicObject = (fileDynamicCollectionDynamicObj this)
        dynamicKvs <- (TypelayerCommonCollections.binaryCollectionSlice dynamicObject offset size)
        TypelayerCommonPrelude.listReduceM dynamicKvs [] $ \reduceDynamicKvsIndex reduceDynamicKvsElem reduceDynamicKvsAcc -> do
            collectionEncoding <- (TypelayerCommonCollections.binaryCollectionEncoding dynamicObject )
            newDecoder <- (TypelayerCommonDynamic.encodingStartDecoding collectionEncoding (TypelayerCommonCollections.binaryKVValue reduceDynamicKvsElem))
            fileValue <- (TypelayerCSchemasWorldDynamic.dynamicGetObjFile newDecoder)
            let appendList0 = ((FileKV{fileKVKey=(TypelayerCommonCollections.binaryKVKey reduceDynamicKvsElem), fileKVValue=fileValue}):reduceDynamicKvsAcc)
            return appendList0

File: typescript/typelayer-c/bin/schemas/world/collections.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// Module: schemas.world.collections
const typelayerCSchemasWorldDynamic = __importStar(require("./dynamic"));
const typelayerCommonCollections = __importStar(require("typelayer-common"));
class FileKV {
    constructor(arg) {
        Object.assign(this, arg);
    }
}
exports.FileKV = FileKV;
class FileDynamicCollection {
    constructor(arg) {
        this.kind = "FileDynamicCollection";
        Object.assign(this, arg);
    }
    fileCollectionDelete(toDelete) {
        return __awaiter(this, void 0, void 0, function* () {
            let dynamicObject;
            dynamicObject = this.dynamicObj;
            return yield dynamicObject.binaryCollectionDelete(toDelete);
        });
    }
    fileCollectionFetch(toFetch) {
        return __awaiter(this, void 0, void 0, function* () {
            let collectionEncoding;
            let dynamicKvs;
            let dynamicObject;
            let fileValue;
            let newDecoder;
            dynamicObject = this.dynamicObj;
            dynamicKvs = yield dynamicObject.binaryCollectionFetch(toFetch);
            let reduceDynamicKvsIndex = 0;
            let reduceDynamicKvsAcc = [];
            for (let reduceDynamicKvsElem of dynamicKvs) {
                collectionEncoding = dynamicObject.binaryCollectionEncoding();
                newDecoder = collectionEncoding.encodingStartDecoding(reduceDynamicKvsElem.value);
                fileValue = typelayerCSchemasWorldDynamic.dynamicGetObjFile(newDecoder);
                reduceDynamicKvsAcc.push(new FileKV({ key: reduceDynamicKvsElem.key, value: fileValue }));
                reduceDynamicKvsAcc = reduceDynamicKvsAcc;
                reduceDynamicKvsIndex++;
            }
            return reduceDynamicKvsAcc;
        });
    }
    fileCollectionSave(toSave) {
        return __awaiter(this, void 0, void 0, function* () {
            let collectionEncoding;
            let dynamicList;
            let dynamicObject;
            let elemBinary;
            let newEncoder;
            dynamicObject = this.dynamicObj;
            let convertLoopIndex = 0;
            let convertLoopAcc = [];
            for (let convertLoopElem of toSave) {
                collectionEncoding = dynamicObject.binaryCollectionEncoding();
                newEncoder = collectionEncoding.encodingStartEncoding();
                typelayerCSchemasWorldDynamic.dynamicPutFile(newEncoder, convertLoopElem.value);
                elemBinary = newEncoder.encoderToBinary();
                convertLoopAcc.push(new typelayerCommonCollections.BinaryKV({
                    key: convertLoopElem.key,
                    value: elemBinary
                }));
                convertLoopAcc = convertLoopAcc;
                convertLoopIndex++;
            }
            dynamicList = convertLoopAcc;
            return yield dynamicObject.binaryCollectionSave(dynamicList);
        });
    }
    fileCollectionSlice(offset, size) {
        return __awaiter(this, void 0, void 0, function* () {
            let collectionEncoding;
            let dynamicKvs;
            let dynamicObject;
            let fileValue;
            let newDecoder;
            dynamicObject = this.dynamicObj;
            dynamicKvs = yield dynamicObject.binaryCollectionSlice(offset, size);
            let reduceDynamicKvsIndex = 0;
            let reduceDynamicKvsAcc = [];
            for (let reduceDynamicKvsElem of dynamicKvs) {
                collectionEncoding = dynamicObject.binaryCollectionEncoding();
                newDecoder = collectionEncoding.encodingStartDecoding(reduceDynamicKvsElem.value);
                fileValue = typelayerCSchemasWorldDynamic.dynamicGetObjFile(newDecoder);
                reduceDynamicKvsAcc.push(new FileKV({ key: reduceDynamicKvsElem.key, value: fileValue }));
                reduceDynamicKvsAcc = reduceDynamicKvsAcc;
                reduceDynamicKvsIndex++;
            }
            return reduceDynamicKvsAcc;
        });
    }
}
exports.FileDynamicCollection = FileDynamicCollection;
